### PR TITLE
Update vgg.py in example/classification

### PR DESCRIPTION
**Pull request by solin319**
Reference to the program in gluon/model_zoo/vision/vgg.py, I update the vgg.py in example/classification.
1. The new vgg include vgg-11,13,16,19. Parameter 'num_layers' means the number of layers for the variant of densenet.
2.  Parameter ‘batch_norm’ controls weather to use  batch normalization.
3. Parameter 'dtype' describe the data precision during the network execution.

@piiswrong

### CODE REVIEW

File: example/image-classification/symbols/vgg.py [hunk 1]
```diff
@@ -20,59 +20,57 @@
...
-    drop6 = mx.symbol.Dropout(data=relu6, p=0.5, name="drop6")
-    # group 7
-    fc7 = mx.symbol.FullyConnected(data=drop6, num_hidden=4096, name="fc7")
-    relu7 = mx.symbol.Activation(data=fc7, act_type="relu", name="relu7")
-    drop7 = mx.symbol.Dropout(data=relu7, p=0.5, name="drop7")
-    # output
-    fc8 = mx.symbol.FullyConnected(data=drop7, num_hidden=num_classes, name="fc8")
-    softmax = mx.symbol.SoftmaxOutput(data=fc8, name='softmax')
-    return softmax
+def get_symbol(num_classes, num_layers, batch_norm=False, dtype='float32', **kwargs):
```
Comment (piiswrong) [inline example/image-classification/symbols/vgg.py]: looks like this is not backward compatible? Give a default value to num_layers?

File: example/image-classification/symbols/vgg.py [hunk 2]
```diff
@@ -20,59 +20,57 @@
...
+        Use batch normalization. 
+    dtype: str, float32 or float16
+        Data precision.   
+    """
+    vgg_spec = {11: ([1, 1, 2, 2, 2], [64, 128, 256, 512, 512]),
+                13: ([2, 2, 2, 2, 2], [64, 128, 256, 512, 512]),
+                16: ([2, 2, 3, 3, 3], [64, 128, 256, 512, 512]),
+                19: ([2, 2, 4, 4, 4], [64, 128, 256, 512, 512])}
+    if not vgg_spec.has_key(num_layers):        
+        raise ValueError("no experiments done on num_layers {}, you can do it yourself".format(num_layers))
```
Comment (piiswrong) [inline example/image-classification/symbols/vgg.py]: "Invalide num_layers. Possible choices are ..."

File: python/mxnet/optimizer.py [hunk 3]
```diff
@@ -383,9 +383,9 @@ def create_state(self, index, weight):
     def update(self, index, weight, grad, state):
         assert(isinstance(weight, NDArray))
         assert(isinstance(grad, NDArray))
+        self._update_count(index)
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: ?

Comment (solin319) [reply python/mxnet/optimizer.py]: If self._update_count(index) was located after self._get_lr(index), the parameter 'self.num_update' used in the function self._get_lr(index) will like bellow.
**INFO:root:index: 0    num_update: 0**
INFO:root:index: 1    num_update: 1
INFO:root:index: 2    num_update: 1
INFO:root:index: 3    num_update: 1
INFO:root:index: 4    num_update: 1
INFO:root:index: 5    num_update: 1
INFO:root:index: 6    num_update: 1
INFO:root:index: 7    num_update: 1
**INFO:root:index: 0    num_update: 1**
INFO:root:index: 1    num_update: 2
INFO:root:index: 2    num_update: 2
INFO:root:index: 3    num_update: 2
INFO:root:index: 4    num_update: 2
INFO:root:index: 5    num_update: 2
INFO:root:index: 6    num_update: 2
INFO:root:index: 7    num_update: 2
**INFO:root:index: 0    num_update: 2**
INFO:root:index: 1    num_update: 3
INFO:root:index: 2    num_update: 3
INFO:root:index: 3    num_update: 3
INFO:root:index: 4    num_update: 3
INFO:root:index: 5    num_update: 3
INFO:root:index: 6    num_update: 3
INFO:root:index: 7    num_update: 3

In the first time update, the 'self.num_update' used in index_0 is 0 and 1 for other index. The same wrong will occur in the next update loop. 
So when lr for indexes except index_0 changed in the function self.lr_scheduler, the index_0 will get the new lr in the next loop.
I move self._update_count before self.get_lr. It can make the lr_scheduler correctly.

**Comment by solin319** [comment]
I have added default num_layers and rewrited 'ValueError'.

@piiswrong

**Comment by piiswrong** [comment]
merged thanks
